Refactor post by Batuhan (#2835)
iterative/dvc.org: content/blog/2021-09-22-refactor.md

negatives (a case where the tool didn't change when it should have), or even
not being able to truly express the structural concepts of the programming
language in them (since many of the programming languages have actual grammars
which are impossible to manifest in regexes).

This is where [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)s shine.
They are the common building blocks of source code; produced by a parser that
actually understands the language's syntax and creates a tree object where
smaller parts (e.g. tokens) are ordered in a way that they are related by their
syntactical meanings.

```python
password = input("password? ")
if password == secrets.get("my_password"):
    print("correct")
else:
    print("incorrect")
```

For example, the AST for the code above will look like this:

![Fundamentals of MLOps](/uploads/images/2021-09-22/ast.png) _Abstract Syntax
Tree_

The top-most "root" node of this tree represents a single Python file. Each file
consists of a number of statements (e.g. function definitions, loops, etc.). For
our example we have only 2 statements: an assignment (to `password`), and an

> `if` statement. Each of these statements in turn have child nodes as defined by

Reviewer: ```suggestion
`if` statement. Each of these statements in turn has child nodes as defined by
```

earlier edited suggestion merged edit :)

Author: statements have, not statements has, no?

Author: Nope. Each has....  got it.

Author: I hated grammar. 😬

﻿---
title: Easy Stuctural Refactors to Python Source Code
date: 2021-09-22
description: |
  Simple, hassle-free, dependency-free, AST based source code refactoring toolkit.

descriptionLong: |
  [Refactor](https://github.com/isidentical/refactor) is a source code 

>   refactoring engine that takes the advantage of the AST module in the standard
>   library. Through using it, we can find-and-fix patterns in massive codebases.

Reviewer: ```suggestion
  refactoring engine. By taking advantage of the Python standard library's
  [AST](https://docs.python.org/3/library/ast.html) module, we can find-and-fix
  patterns in massive codebases.
```

picture: 2021-09-22/ast.png
author: batuhan_taskaya
commentsUrl: https://discuss.dvc.org/t/easy-structural-refactors-to-python-source-code/895
tags:

>   - Refactor

Reviewer 2: can we please introduce `engineering` and `open source` (or `open-source`?) tags? Probably propagate them to all things that were written by the eng team - Peter, Casper

Reviewer: in this PR or in a follow-up?

Author: Adding tags to this post.  Let's add the tags to the others separately pls.

  - AST
---

> It is common for bigger codebases to grow to a complexity where it is really
> hard for someone to go and refactor stuff manually on every occurence. There are
> automated solutions for this, like regular expression based search and replace
> tools, but the main problem with those is that they treat source code like a
> regular text document, which might lead to some issues like increased number of
> some false positives (a case where the tool did change when it shouldn't), true
> negatives (a case where the tool didn't change when it should have), or even
> not being able to truly express the structural concepts of the programming
> language in them (since many of the programming languages have actual grammars
> which are impossible to manifest in regexes).

Reviewer: ```suggestion
It is common for big codebases to grow to a complexity where it nearly
impossible for someone to tediously and flawlessly refactor things manually
everywhere. The main problem with existing automated solutions (such as
regex-based find-and-replace tools) is that they treat source code like a plain
text document. This often results in false positives (tools making changes when
they shouldn't), true negatives (not changing what they should). This is
primarily due to a lack of ability to truly encapsulate structural concepts of
the programming language: syntax and grammar that are impossible to manifest in
regexes.
```

Author: *is* nearly impossible.  Needs a conjunction (or) in "This often..."  I'll correct and send up with the new tags.  👍 👍 to other changes @casperdcl!  Thank you!

Reviewer: I just re-read and edited a few comments, hope you pick up changes.

Author: Ok.  @casperdcl should it be and/or?

Author: Do these things always happen in conjunction?  (and) It doesn't seem so the way @isidentical wrote it, but I don't understand enough to know.

Reviewer: "and" makes more sense ("This [state of affairs] often results in [problem X] and [Y]" rather than "This often results [in problem] [X] or [Y]")

Reviewer: > Do these things always happen in conjunction?

identical sentiment to "telling lies leads to unhappiness and/or distrust." Both work; nobody's going to notice difference between `AND` vs `OR` vs `XOR` logic :)

understands the language's syntax, and creates a tree object where smaller parts
(e.g. tokens) are ordered in a way that they are related in terms of their
syntactical meanings.

```python
password = input("password? ")
if password == secrets.get("my_password"):
    print("correct")
else:
    print("incorrect")
```

For example, the AST for the code above will look like this:

![Fundamentals of MLOps](/uploads/images/2021-09-22/ast.png) _Abstract Syntax
Tree_

It is in the shape of a tree, where the root point represents a single Python
file. Each file consists from a number of statements (e.g function definitions,
loops), and for our example we have only 2 statements: an assignment (to
password), and an if statement. Each of these statements have their child
nodes/leaves, which are all defined in the
[Python's ASDL](https://docs.python.org/3/library/ast.html#abstract-grammar).

## Refactoring source code through ASTs

What [refactor](https://github.com/isidentical/refactor) does is that, it
enables the possibility of simply matching ASTs, and then transforming them to
different ones with doing as little as possible in terms of touching the other
parts of the source code.

> 

Reviewer: ```suggestion

For example, consider this code:

```

```python
foo = [
    1,
    2
]

foo_2 = ['a', *foo]

if foo[0] >= 1:
    assert secrets.get("foo") == foo
```

> As a simple demo, let's try to find and replace all usages of `foo` variable
> name with `bar`, but without actually changing things inside strings or partial
> matches like `foo_2`.

Reviewer: ```suggestion
As a simple example, let's try to find and replace all instances of the `foo`
variable with `bar`... but without changing things inside strings or partial
matches like `foo_2`.
```

```python
import ast
import refactor
```

The first thing we need to do is define a rule. Each rule is a class that
defines a single entrypoint (`match())`), takes AST nodes from the tree, and
either rejects them (via raising an `AssertionError` or just returning `None`)
or accepts them (via returning a `refactor.Action`).

```python
class ReplaceFoo(refactor.Rule):

    def match(self, node):
```

> What we do in the `match()` method is simple, we will look for all `Name`s
> (which is what the actual identifier is wrapped when used in the contexts
> above), and check whether the `node.id` is `foo` or not.

Reviewer: ```suggestion
Next, in the `match()` method, we will look for all `Name`s (which is what the
actual identifier is wrapped in), and check whether its `id` is `foo`.
```

```python
        assert isinstance(node, ast.Name)
        assert node.id == "foo"
```

If any of these assertions fail, the function will terminate at that point and
the engine will move to the next node in the tree. But if we succeed, then it
means that we have a match, and we should return some sort of an action. The
simplest thing we can use is just returning a `refactor.ReplacementAction` which
takes this node and replaces it with the given argument.

```python
        return refactor.ReplacementAction(
            node,
            ast.Name("bar", node.ctx)
        )
```

> And this is it. If we want to try it, we can simply create a CLI application out
> of our refactoring rules via `refactor.run()`.

Reviewer: ```suggestion
And that's it! To run this refactoring, we can simply create a CLI application
from our rules via `refactor.run()`:
```

```python
if __name__ == "__main__":
    refactor.run(rules=[ReplaceFoo])
```

> And if we run it on the file above, we will get this diff:

Reviewer: ```suggestion
If we run it on the file above, we will get this `diff`:
```

```diff
@@ -1,9 +1,9 @@

> 

Reviewer: ```suggestion
```

remove extra line

-foo = [
+bar = [
     1,
     2
 ]

-foo_2 = ['a', *foo]
+foo_2 = ['a', *bar]

-if foo[0] >= 1:
-    assert secrets.get("foo") == foo
+if bar[0] >= 1:
+    assert secrets.get("foo") == bar
```

> Which indicates all of the usages against `foo` is replaced, but items like
> `foo_2` or `"foo"` is still there as we have expected.

Reviewer: ```suggestion
All instances of the `foo` variable have been replaced, but items like `foo_2`
and `"foo"` are left alone as expected!
```

## Going Deeper

> Obviously not all refactorings are as easy as this, so refactor is equipped with
> more features like different actions, and more advanced features like observers
> and representatives for context manager. If you are curious about them, be sure
> to check out it's [documentation!](https://refactor.readthedocs.io/en/latest/)

Reviewer: ```suggestion
Obviously not all refactorings are as simple as this, so `refactor` is equipped
with more features like different actions, observers and representatives for
context manager. If you are curious about these and more advance features, be
sure to check out the
[`refactor` documentation](https://refactor.readthedocs.io/en/latest)!
```

Author: will correct locally with other changes ->advanced

Reviewer: fml today not my day for typos. edited.

---

_Do you have any use case questions or need support? Join us in
[Discord](https://discord.com/invite/dvwXA2N)!_

_Head to the [DVC Forum](https://discuss.dvc.org/) to discuss your ideas and

> best practices._

Reviewer: is this relevant?

Author: Good point.  No, not really. Will drop.

  [AST](https://docs.python.org/3/library/ast.html) module, we can find-and-fix
  patterns in massive codebases.

> picture: 2021-09-22/refactor.png

Reviewer: ```suggestion
picture: 2021-09-24/refactor.png
```

Reviewer: 460dac8
